osu.Game/Rulesets/Difficulty/PerformanceBreakdownCalculator.cs

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+using osu.Framework.Extensions;
+using osu.Game.Beatmaps;
+using osu.Game.Rulesets.Mods;
+using osu.Game.Rulesets.Scoring;
+using osu.Game.Scoring;
+
+namespace osu.Game.Rulesets.Difficulty
+{
+    public class PerformanceBreakdownCalculator
+    {
+        private readonly BeatmapManager beatmapManager;
+        private readonly BeatmapDifficultyCache difficultyCache;
+        private readonly ScorePerformanceCache performanceCache;
+
+        public PerformanceBreakdownCalculator(BeatmapManager beatmapManager, BeatmapDifficultyCache difficultyCache, ScorePerformanceCache performanceCache)
+        {
+            this.beatmapManager = beatmapManager;
+            this.difficultyCache = difficultyCache;
+            this.performanceCache = performanceCache;
+        }
+
+        [ItemCanBeNull]
+        public async Task<PerformanceBreakdown> CalculateAsync(ScoreInfo score, CancellationToken cancellationToken = default)
+        {
+            PerformanceAttributes performance = await performanceCache.CalculatePerformanceAsync(score, cancellationToken).ConfigureAwait(false);
+
+            ScoreInfo perfectScore = await getPerfectScore(score, cancellationToken).ConfigureAwait(false);
+            if (perfectScore == null)
+                return null;
+
+            PerformanceAttributes perfectPerformance = await performanceCache.CalculatePerformanceAsync(perfectScore, cancellationToken).ConfigureAwait(false);
+
+            return new PerformanceBreakdown { Performance = performance, PerfectPerformance = perfectPerformance };
+        }
+
+        [ItemCanBeNull]
+        private Task<ScoreInfo> getPerfectScore(ScoreInfo score, CancellationToken cancellationToken = default)
+        {
+            return Task.Factory.StartNew(() =>
+            {
+                IBeatmap beatmap = beatmapManager.GetWorkingBeatmap(score.BeatmapInfo).GetPlayableBeatmap(score.Ruleset, score.Mods);
+                ScoreInfo perfectPlay = score.DeepClone();
+                perfectPlay.Accuracy = 1;
+                perfectPlay.Passed = true;
+
+                // calculate max combo
+                var difficulty = difficultyCache.GetDifficultyAsync(
+                    beatmap.BeatmapInfo,
+                    score.Ruleset,
+                    score.Mods,
+                    cancellationToken
+                ).GetResultSafely();
```

Comment:
Why is this using a blocking call rather than awaits? Also how does `GetResultSafely()` not crash here? `Task.Factory.StartNew()` should queue up the work inside onto a TPL thread, and this `GetResultSafely()` will block on an operation that's potentially taking place on another pool thread, possibly triggering a deadlock? I'd understand if the `Task.Factory.StartNew()` call specified `LongRunning` but it doesn't...?

The only explanation I have right now is that the result always already happens to be available because of prior computations (on song select or otherwise). Either way I'd say this should use a proper async flow.